pytorch_lightning/utilities/auto_restart.py

```diff
@@ -207,29 +207,189 @@ def __iter__(self) -> Iterator:
+    workers and call the reset function.
+
+    Returns:
+        iterator: Return the iterator generated from the provided ``DataLoader``.
+    """
+    iter_dataloader = iter(dataloader)
+    num_workers = getattr(iter_dataloader, "_num_workers", 0)
+    if state_dict["num_workers"] != num_workers:
+        raise MisconfigurationException(
+            f"The provided ``num_workers`` {num_workers} doesn't match the one used "
+            f"while generating the checkpoint: {state_dict['num_workers']}"
+        )
+    if isinstance(iter_dataloader, _MultiProcessingDataLoaderIter):
+        # move back to 0
+        while next(iter_dataloader._worker_queue_idx_cycle) != 0:
+            pass
+        # increment previous worker
+        if isinstance(state_dict["previous_worker"], int):
+            for _ in range(state_dict["previous_worker"] - 1):
+                next(iter_dataloader._worker_queue_idx_cycle)
+        iter_dataloader._reset = iter_dataloader._original_reset
+        iter_dataloader._reset(dataloader, first_iter=True)
+    return iter_dataloader
+
+
+def dataloader_to_state_dict(
+    dataloader: DataLoader,
+    iterator: Iterator,
+    num_batches_processed: int = None,
+) -> List[Dict[str, Any]]:
+    """
+    Convert a dataloader to its associated state dict
+    """
+    out = {}
+    if iterator is not None:
+        out.update(find_current_worker(iterator))
+
+    if not isinstance(dataloader.dataset, CaptureIterableDataset):
+        fast_forward_sampler = find_fast_forward_samplers(dataloader)
+        if fast_forward_sampler is not None:
+            out.update(fast_forward_sampler.state_dict(num_batches_processed=num_batches_processed))
+    return out
+
+
+def dataloader_load_state_dict(dataloader: DataLoader, state_dict: List[Dict[str, Any]]) -> DataLoader:
+    """
+    Reload ``DataLoader`` fast-forward sampler state dict.
+    """
+    fast_forward_sampler = find_fast_forward_samplers(dataloader)
+
+    if isinstance(fast_forward_sampler, Sampler):
+        state_dict = {k: v for k, v in state_dict.items() if k not in ("num_workers", "previous_worker")}
+        fast_forward_sampler.load_state_dict(state_dict)
+
+    return dataloader
+
+
+def find_current_worker(iterator: Iterator) -> Dict[str, Optional[int]]:
+    """
+    Find the current DataLoader Iterator worker if multiple workers were used..
+    """
+    num_workers = getattr(iterator, "_num_workers", 0)
+    if isinstance(iterator, _MultiProcessingDataLoaderIter):
+        next_worker = (next(iterator._worker_queue_idx_cycle)) % num_workers
+        previous_worker = (next_worker - 1) % num_workers
+        while next(iterator._worker_queue_idx_cycle) != previous_worker:
+            pass
+    else:
+        previous_worker = None
+
+    return {"num_workers": num_workers, "previous_worker": previous_worker}
+
+
+def sampler_metadata_collate(samples: List, dataset: Dataset, default_collate: Callable) -> Dict:
```

Comment:
Most if not all of these new functions should be made protected

Reply:
Done !